### ISSUE TITLE
Support for parallel and serial execution of set of queries

### ISSUE THREAD

**Issue by Harkishen-Singh** [zairza-cetb/bench-routes#239]
The current querier implementation is healthy enough for running thousands of queries in parallel as a set. This support should be provided natively in the `querier` package in `tsdb`. Also, even though serial execution is no extra pain, it will be great to have them as well. 

The basic implementation bing, a register function to register a query in a set (unique), which can be set as `executionTypeParallel` or `executionTypeParallel`. Then, the `querySet.Run()` triggers the querying processes.

**Comment by aquibbaig** [comment]
I would work on this issue. It would help me understand querier properly.

**Comment by aquibbaig** [comment]
> The current querier implementation is healthy enough for running thousands _(possibly millions if the ulimit is increased)_ of queries in parallel as a set. This should be benchmarked as well so as to keep a track of how well and efficient our querier code is. This support should be provided natively in the `querier` package in `tsdb`. Also, even though serial execution is no extra pain, it will be great to have them as well.
> 
> The basic implementation bing, a register function to register a query in a set (unique), which can be set as `executionTypeParallel` or `executionTypeParallel`. Then, the `querySet.Run()` triggers the querying processes.

@Harkishen-Singh  You mean to say `executionTypeParallel` or `executionTypeSerial` ?

**Comment by Harkishen-Singh** [comment]
@aquibbaig I think right now you aren't working on this. So I others are open to taking this up.

**Comment by geforce6t** [comment]
@Harkishen-Singh i would like to work on this issue.

**Comment by Harkishen-Singh** [comment]
Sure @geforce6t. Please go ahead.

**Comment by devansh42** [comment]
Hi! @Harkishen-Singh if @geforce6t is not working on this project, I'd like to proceed on this. But I've got a small doubt that, what do you mean by Parallel execution, does it mean concurrent execution of queries or running them parallel on a multicore CPU in real. Because in case of real parallel execution, We would require to add some parallel computing library.

**Comment by geforce6t** [comment]
@devansh42 sorry for not mentioning before, I have some obligations, so i am not currently working on this you can take up this issue.

**Comment by Tushar3099** [comment]
@devansh42 I think `Go` itself provides go-routines, channels, and wait-groups for parallel execution; there is no need for other parallel computing libraries.
Please correct me if I am wrong.

**Comment by devansh42** [comment]
@Tushar3099 you are right, Go provides go-routines, channels are wait groups, etc but these are concurrency primitives. There is a subtle difference between concurrency as You can dance while signing, this is called parallelism. But think about talking while chewing your food, this is called concurrency means handling multiple tasks at once.
A set of concurrent tasks might run in parallel eventually but it is not guaranteed, think of a situation when two goroutines are residing on the same thread on a dual-core system, in that case, concurrency is achieved but not parallelism.
There is an amazing talk on [Concurrency Vs Parallelism](https://www.youtube.com/watch?v=oV9rvDllKEg) by Rob Pike.
So to achieve real parallelism, we need to have certain kinds of libraries.

**Comment by Tushar3099** [comment]
@devansh42 Thanks for the descriptive and clear explanation of concurrency vs parallelism; that point is right. But don't we get runtime.GOMAXPROCS() (which is set by default, no. of cores in the user computer) in GoLang that enables Go to execute parallelly on different threads. I know other library built solely for this purpose will do things more efficiently, but it's not like go executes everything concurrently and not parallelly. That depends on the no. of cores allocated to the program.
Even the talk that you suggested by Rob Pike,  says we can make a concurrent model run parallelly if they run on different threads. 
Thanks for the talk recommendation. That was Great!!

**Comment by Harkishen-Singh** [comment]
Sorry, we mean concurrency here. Go scheduler is a good guy to schedule goroutines and we can trust on him. Third party libraries dependency is something I don't think we require at the moment. 

@devansh42 @Tushar3099 thank you for your thoughts here.

**Comment by Tushar3099** [comment]
@devansh42 Are you working on this issue or I can take this up?

**Comment by devansh42** [comment]
Sure @Tushar3099, right now I'm working on another issue. So you can proceed.

**Comment by devansh42** [comment]
Hey @Tushar3099, I found this issue quite interesting, Can we collaborate on this.

**Comment by devansh42** [comment]
> @devansh42 Thanks for the descriptive and clear explanation of concurrency vs parallelism; that point is right. But don't we get runtime.GOMAXPROCS() (which is set by default, no. of cores in the user computer) in GoLang that enables Go to execute parallelly on different threads. I know other library built solely for this purpose will do things more efficiently, but it's not like go executes everything concurrently and not parallelly. That depends on the no. of cores allocated to the program.
> Even the talk that you suggested by Rob Pike, says we can make a concurrent model run parallelly if they run on different threads.
> Thanks for the talk recommendation. That was Great!!

* As Rob Pike said in that talk, Parallelism is a subset of Concurrency i.e. If you are executing two tasks in parallel, we can say that you are going them concurrently.
* Parallelism is an Eventual thing, It may happen but is not guaranteed.
* **GOMAXPROCS** just limits the number of Logical CPU core, it won't set some kind of degree of parallelism.
I hope this will help you, please correct me If I'm wrong somewhere.

**Comment by Tushar3099** [comment]
@devansh42 Quiet happy to see a healthy discussion, but I don't think this is the right platform to do so as this may push the main issue to the side. Even though I would like to give my remarks to your statement. IMHO : 
- Of course, you are right, and by all means, I also tried to convey the same thing. Concurrency is a model so that if it is scaled and run on multiple cores, the same model will have improved performance, and if it is executed on a single-core, it will have the same performance as non-concurrent code.
- As I said earlier: 

> I know other library built solely for this purpose will do things more efficiently, but it's not like go executes everything concurrently and not parallelly. That depends on the no. of cores allocated to the program.

- Also : 

> runtime.GOMAXPROCS() (which is set by default, no. of cores in the user computer) in GoLang

I was telling the same thing that you said afterward. It limits the no. of the CPU core, which would eventually lead to running concurrent code on multiple cores, which, according to what we discussed, eventually runs parallelly.

I am sorry if some of my explanations weren't correct. I am new to Go and learning things right now. I would hence like to end the discussion on Parallel vs. Concurrency as discussing the main issue will be more fruitful. I sure did learn a lot from the discussion.

**Comment by Tushar3099** [comment]
@devansh42 I would have surely loved to collaborate with you, but I want to work on this issue solely as I may not catch up to your pace. I hope you understand.

**Comment by Tushar3099** [comment]
@Harkishen-Singh @aquibbaig @muskankhedia  I was working on QueryBatch data struct which will execute the batch querying. I came up with : 
```
      type QueryBatch struct {
	      stamp         time.Time
	      querySet	  []*Query
	      Type		  uint8
	      queryResponse []QueryResponse
      }
```
But I was wondering how we can map the Query and QueryResponse. Even though we have the queryResponse field in Query, we are not saving the final response ( base ) to q.queryResponse. I was not sure if it was done intentionally for something I may have skipped.
I assume the stamp is for a unique id as we can create one query at a time, similarly, I added a stamp field in the QueryBatch. I thought of a solution : 
- Instead of having a different array of queryResponse, we can save the query response to each Query's queryResponse field.

What are your views on this? Please guide with the right approach.

**Comment by Tushar3099** [comment]
@Harkishen-Singh, Can you please give suggestions on the above approach so that I can raise a PR for the issue?

**Comment by Tushar3099** [comment]
@aquibbaig @Harkishen-Singh please give your suggestions on the above approach. If the approach is correct then I can raise a PR on this issue.
